### ISSUE TITLE
crash after dismissing a presented UICollectionViewController where InfiniteCarousel is the controller's collection view

### ISSUE THREAD

**Issue by klundberg** [willowtreeapps/ouroboros#16]
If you create a collection view controller where the InfiniteCarousel is set to be the collection view of said controller, and you then present/scroll/dismiss that view controller, the next time you try to focus something new, the app will crash at this line:

https://github.com/willowtreeapps/ouroboros/blob/develop/Ouroboros/InfiniteCarousel.swift#L92

The problem is that the focus engine seems to keep a reference to the Infinite carousel, and when you refocus, that reference is removed. UIScrollView's dealloc method manually sets datasource and delegate to nil, and that set here causes the overridden property for delegate to trigger the set, which tries to re-set super.delegate. Doing this causes a trap, since the view is being deallocated it is not allowed to be set to a weak reference which it is doing by setting super.delegate to self.

**Comment by ianterrell** [comment]
Thank you for the report. Definitely sounds like a bug.

If you have a small sample project that demonstrates this issue it could definitely help speed up a fix. PRs also welcome of course. :)

**Comment by snofla** [comment]
I fixed it by making deinit set a variable, and check in the offending lines for this variable. If variable is set I call the super implementation of the setter. HTH.

**Comment by klundberg** [comment]
I haven't needed to use this recently, it was for a POC, but if I have time this weekend I'll try to make a sample proj for it and see if I can reproduce it there.

**Comment by ehynds** [comment]
@snofla mind sharing code for how you fixed this? I'm running into the issue as well.

**Comment by ianterrell** [comment]
It sounds to me like @snofla was doing something like this:

```swift
ViewController {
  deinit {
    infiniteCarousel.deiniting = true
  }
}

InfiniteCarousel {
  var deiniting = false
  set {
    guard !deiniting else {
      super.delegate = newValue
      return
    }

    rootDelegate = newValue
    super.delegate = self
  }
}
```

In my mind the `nil` case might be the most relevant here... I'm not sure. If so, something like this could work:
```swift
set {
  guard let delegate = newValue else {
    rootDelegate = nil
    super.delegate = nil
    return`
  }

  rootDelegate = delegate
  super.delegate = self
}
```

We haven't run into this case yet and I've been too busy to repro it manually; if you attach a repro and a PR I will review and try to get it merged into a new release.

**Comment by snofla** [comment]
@ianterrell Yes, the former, I'll try to come up with a reproducible scenario.

**Comment by BProg** [comment]
I have resolved this bug on my fork. But a strange thing is that if I use InfiniteCarousel as framework, it still crashes, but when adding file directly to the project, it does not crash with the fix.
